algebra/ring.lean

```diff
@@ -15,6 +15,85 @@ section
   (left_distrib n 1 1).trans (by simp)
 end
 
+section units
+  variables [semiring α]
+  variable (α)
+
+  structure invertible :=
+    (val : α)
+    (inv : α)
+    (property : val * inv = 1 ∧ inv * val = 1)
+
+  variables {a b c : invertible α}
+  variable {α}
+
+  def ext (HAB : a.val = b.val) : a = b :=
+  begin
+    cases a,
+    cases b,
+    dsimp at HAB,
+    congr,
+    exact HAB,
+    exact calc
+        inv = inv * (val_1 * inv_1) : by rw [property_1.1, mul_one]
+        ... = (inv * val_1) * inv_1 : by rw [mul_assoc]
+        ... = (inv * val) * inv_1   : by rw [HAB]
+        ... = inv_1                 : by rw [property.2, one_mul]
+  end
+
+  lemma mul_four {a b c d : α} :
```

Comment:
This can go in `group.lean`, let's say `mul_assoc_four`. (The name may need to change later, but this should do for now.)

Reply:
I don't like these kind of lemmas. Is it really often used? Why not directly calling the simplifier?